File: jsrc/kernelprolog/io/SxxMachine/ClauseWriter.java
package SxxMachine;

import SxxMachine.pterm.Expect;
import SxxMachine.pterm.TermData;

/**
 * Writer
 */
public class ClauseWriter extends CharWriter {
    public ClauseWriter(String f, Prog p) {
        super(f, p);
    }

    public ClauseWriter(Prog p) {
        super(p);
    }

    @Override
    public int putElement(Term t) {
        if (null == writer)
            return 0;
        String s = null;
        if ((t.isCompound()) && "$string".equals(Expect.asStruct(t).fname())) {
            Term Xs = Expect.asConst(Expect.asStruct(t).ArgDeRef(0));
            s = TermData.charsToString(Xs);
        } else {
            s = t.pprint();
        }
        IO.print(writer, s);
        return 1;
    }
}

File: src/kernelprolog/io/SxxMachine/ClauseWriter.cpp
using namespace std;

#include "ClauseWriter.h"
#include "../../terms/SxxMachine/Prog.h"
#include "../../../machine/SxxMachine/Term.h"
#include "../../terms/SxxMachine/Expect.h"
#include "../../../machine/SxxMachine/SymbolTerm.h"
#include "IO.h"

namespace SxxMachine
{

	ClauseWriter::ClauseWriter(const wstring& f, Prog* p) : CharWriter(f, p)
	{
	}

	ClauseWriter::ClauseWriter(Prog* p) : CharWriter(p)
	{
	}

	int ClauseWriter::putElement(Term* t)
	{
		if(nullptr == writer)
		{
			return 0;
		}
		wstring s = "";
		if((t->isStructure()) && "$string" == Expect::asStruct(t)->name())
		{
			SymbolTerm* Xs = Expect::asConst(Expect::asStruct(t)->ArgDeRef(0));
			s = Term::charsToString(Xs);
		} else
		{
			s = t->pprint();
		}
		IO::print(writer, s);
		return 1;
	}
}
